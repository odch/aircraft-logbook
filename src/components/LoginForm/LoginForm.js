import React from 'react'
import PropTypes from 'prop-types'
import Button from '@material-ui/core/Button'
import FormControl from '@material-ui/core/FormControl'
import Input from '@material-ui/core/Input'
import InputLabel from '@material-ui/core/InputLabel'
import Typography from '@material-ui/core/Typography'
import withStyles from '@material-ui/core/styles/withStyles'

const styles = theme => ({
  layout: {
    width: 'auto',
    display: 'block',
    marginTop: theme.spacing.unit * 8,
    marginLeft: theme.spacing.unit * 3,
    marginRight: theme.spacing.unit * 3,
    [theme.breakpoints.up(400 + theme.spacing.unit * 3 * 2)]: {
      width: 400,
      marginLeft: 'auto',
      marginRight: 'auto'
    }
  },
  failureMessage: {
    color: theme.palette.error.dark
  }
})

class LoginForm extends React.Component {
  handleUsernameChange = e => {
    this.props.setUsername(e.target.value)
  }

  handlePasswordChange = e => {
    this.props.setPassword(e.target.value)
  }

  handleSubmit = e => {
    e.preventDefault()
    const { username, password } = this.props.loginForm
    this.props.login(username, password)
  }

  render() {
    const {
      classes,
      loginForm: { username, password, failed, submitted }
    } = this.props

    return (
      <React.Fragment>
        <main className={classes.layout}>
          {failed && (
            <Typography className={classes.failureMessage}>
              Falsche Login-Daten
            </Typography>
          )}
          <form onSubmit={this.handleSubmit}>
            <FormControl margin="normal" required fullWidth>
              <InputLabel htmlFor="email">E-Mail</InputLabel>
              <Input
                id="email"
                name="email"
                autoComplete="email"
                onChange={this.handleUsernameChange}
                value={username}
                disabled={submitted}
                data-cy="email"
                autoFocus
              />
            </FormControl>
            <FormControl margin="normal" required fullWidth>
              <InputLabel htmlFor="password">Passwort</InputLabel>
              <Input
                name="password"
                type="password"
                id="password"
                autoComplete="current-password"
                onChange={this.handlePasswordChange}
                value={password}
                disabled={submitted}
                data-cy="password"
              />
            </FormControl>
            <Button
              type="submit"
              fullWidth
              variant="raised"
              color="primary"
              disabled={submitted}
              data-cy="submit"
            >
              Anmelden
            </Button>
          </form>
        </main>
      </React.Fragment>
    )
  }
}

LoginForm.propTypes = {
  loginForm: PropTypes.shape({
    username: PropTypes.string.isRequired,
    password: PropTypes.string.isRequired,
    failed: PropTypes.bool.isRequired,
    submitted: PropTypes.bool.isRequired
  }).isRequired,
  setUsername: PropTypes.func.isRequired,
  setPassword: PropTypes.func.isRequired,
  login: PropTypes.func.isRequired,
  classes: PropTypes.object.isRequired
}

export default withStyles(styles)(LoginForm)
