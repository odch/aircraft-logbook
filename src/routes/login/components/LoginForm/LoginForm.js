import React from 'react'
import PropTypes from 'prop-types'
import { FormattedMessage } from 'react-intl'
import Button from '@material-ui/core/Button'
import FormControl from '@material-ui/core/FormControl'
import Input from '@material-ui/core/Input'
import InputLabel from '@material-ui/core/InputLabel'
import Typography from '@material-ui/core/Typography'
import withStyles from '@material-ui/core/styles/withStyles'

const styles = theme => ({
  failureMessage: {
    color: theme.palette.error.dark
  }
})

class LoginForm extends React.Component {
  handleUsernameChange = e => {
    this.props.setUsername(e.target.value)
  }

  handlePasswordChange = e => {
    this.props.setPassword(e.target.value)
  }

  handleSubmit = e => {
    e.preventDefault()
    const { username, password } = this.props.loginForm
    this.props.login(username, password)
  }

  render() {
    const {
      classes,
      loginForm: { username, password, failed, submitted }
    } = this.props

    return (
      <React.Fragment>
        {failed && (
          <Typography className={classes.failureMessage}>
            <FormattedMessage id="login.failed" />
          </Typography>
        )}
        <form onSubmit={this.handleSubmit}>
          <FormControl margin="normal" required fullWidth>
            <InputLabel htmlFor="email">
              <FormattedMessage id="login.email" />
            </InputLabel>
            <Input
              id="email"
              name="email"
              autoComplete="email"
              onChange={this.handleUsernameChange}
              value={username}
              disabled={submitted}
              data-cy="email"
              autoFocus
            />
          </FormControl>
          <FormControl margin="normal" required fullWidth>
            <InputLabel htmlFor="password">
              <FormattedMessage id="login.password" />
            </InputLabel>
            <Input
              name="password"
              type="password"
              id="password"
              autoComplete="current-password"
              onChange={this.handlePasswordChange}
              value={password}
              disabled={submitted}
              data-cy="password"
            />
          </FormControl>
          <Button
            type="submit"
            fullWidth
            variant="raised"
            color="primary"
            disabled={submitted}
            data-cy="submit"
          >
            <FormattedMessage id="login.login" />
          </Button>
        </form>
      </React.Fragment>
    )
  }
}

LoginForm.propTypes = {
  loginForm: PropTypes.shape({
    username: PropTypes.string.isRequired,
    password: PropTypes.string.isRequired,
    failed: PropTypes.bool.isRequired,
    submitted: PropTypes.bool.isRequired
  }).isRequired,
  setUsername: PropTypes.func.isRequired,
  setPassword: PropTypes.func.isRequired,
  login: PropTypes.func.isRequired,
  classes: PropTypes.object.isRequired
}

export default withStyles(styles)(LoginForm)
