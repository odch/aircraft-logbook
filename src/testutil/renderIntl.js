import React from 'react'
import renderer from 'react-test-renderer'
import { createMount } from '@material-ui/core/test-utils'
import { IntlProvider } from 'react-intl'
import pretty from 'pretty'
import messages from '../messages'

const LOCALE = 'de'

const renderInIntlProvider = component => (
  <IntlProvider locale={LOCALE} messages={messages[LOCALE]}>
    {component}
  </IntlProvider>
)

const renderIntl = (component, options) =>
  renderer.create(renderInIntlProvider(component), options)

export const renderIntlUpdate = (renderer, component, options) =>
  renderer.update(renderInIntlProvider(component), options)

export const renderIntlMaterial = (component, returnPrettyHtml = false) => {
  const rendered = createMount()(renderInIntlProvider(component))
  return returnPrettyHtml ? pretty(rendered.html()) : rendered
}

export default renderIntl
