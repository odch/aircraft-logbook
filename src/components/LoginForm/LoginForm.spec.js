import React from 'react'
import Input from '@material-ui/core/Input'
import renderIntl from '../../testutil/renderIntl'
import LoginForm from './LoginForm'

function renderLoginForm(
  loginForm = { username: '', password: '', failed: false, submitted: false },
  setUsername = () => {},
  setPassword = () => {},
  login = () => {}
) {
  return renderIntl(
    <LoginForm
      loginForm={loginForm}
      setUsername={setUsername}
      setPassword={setPassword}
      login={login}
    />
  )
}

describe('components', () => {
  describe('LoginForm', () => {
    it('renders correctly', () => {
      const loginFormData = {
        username: 'test@example.com',
        password: 'mypassword',
        failed: false,
        submitted: false
      }
      const tree = renderLoginForm(loginFormData).toJSON()
      expect(tree).toMatchSnapshot()
    })

    it('renders error message if login failed', () => {
      const loginFormData = {
        username: 'test@example.com',
        password: '',
        failed: true,
        submitted: false
      }
      const tree = renderLoginForm(loginFormData).toJSON()
      expect(tree).toMatchSnapshot()
    })

    it('calls setUsername function', () => {
      const setUsername = jest.fn()

      const component = renderLoginForm(undefined, setUsername)

      const usernameInput = component.root.find(
        el => el.type === Input && el.props.name === 'email'
      )
      usernameInput.props.onChange({
        target: { value: 'test@example.com' }
      })
      expect(setUsername).toHaveBeenCalledWith('test@example.com')
    })

    it('calls setPassword function', () => {
      const setPassword = jest.fn()

      const component = renderLoginForm(undefined, undefined, setPassword)

      const usernameInput = component.root.find(
        el => el.type === Input && el.props.name === 'password'
      )
      usernameInput.props.onChange({
        target: { value: 'mytestpassword' }
      })
      expect(setPassword).toHaveBeenCalledWith('mytestpassword')
    })

    it('calls login function on submit', () => {
      const loginFormData = {
        username: 'test@example.com',
        password: 'mypassword',
        failed: false,
        submitted: false
      }
      const login = jest.fn()
      const e = {
        preventDefault: jest.fn().mockReturnThis()
      }

      const component = renderLoginForm(
        loginFormData,
        undefined,
        undefined,
        login
      )

      const form = component.root.find(el => el.type === 'form')

      form.props.onSubmit(e)

      expect(e.preventDefault).toHaveBeenCalled()
      expect(login).toHaveBeenCalledWith('test@example.com', 'mypassword')
    })
  })
})
