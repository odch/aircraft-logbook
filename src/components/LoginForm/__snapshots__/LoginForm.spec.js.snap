// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components LoginForm renders correctly 1`] = `
<main
  className="LoginForm-layout-1"
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="MuiFormControl-root-3 MuiFormControl-marginNormal-4 MuiFormControl-fullWidth-6"
    >
      <label
        className="MuiFormLabel-root-14 MuiFormLabel-filled-18 MuiFormLabel-required-19 MuiInputLabel-root-7 MuiInputLabel-formControl-8 MuiInputLabel-animated-11 MuiInputLabel-shrink-10"
        data-shrink={true}
        htmlFor="email"
      >
        E-Mail
        <span
          className="MuiFormLabel-asterisk-20"
        >
           *
        </span>
      </label>
      <div
        className="MuiInputBase-root-34 MuiInput-root-21 MuiInput-underline-25 MuiInputBase-formControl-35 MuiInput-formControl-22"
        data-cy="email"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoComplete="email"
          autoFocus={true}
          className="MuiInputBase-input-44 MuiInput-input-29"
          disabled={false}
          id="email"
          name="email"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={true}
          type="text"
          value="test@example.com"
        />
      </div>
    </div>
    <div
      className="MuiFormControl-root-3 MuiFormControl-marginNormal-4 MuiFormControl-fullWidth-6"
    >
      <label
        className="MuiFormLabel-root-14 MuiFormLabel-filled-18 MuiFormLabel-required-19 MuiInputLabel-root-7 MuiInputLabel-formControl-8 MuiInputLabel-animated-11 MuiInputLabel-shrink-10"
        data-shrink={true}
        htmlFor="password"
      >
        Passwort
        <span
          className="MuiFormLabel-asterisk-20"
        >
           *
        </span>
      </label>
      <div
        className="MuiInputBase-root-34 MuiInput-root-21 MuiInput-underline-25 MuiInputBase-formControl-35 MuiInput-formControl-22"
        data-cy="password"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoComplete="current-password"
          className="MuiInputBase-input-44 MuiInput-input-29 MuiInputBase-inputType-47 MuiInput-inputType-32"
          disabled={false}
          id="password"
          name="password"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={true}
          type="password"
          value="mypassword"
        />
      </div>
    </div>
    <button
      className="MuiButtonBase-root-77 MuiButton-root-51 MuiButton-contained-62 MuiButton-containedPrimary-63 MuiButton-raised-65 MuiButton-raisedPrimary-66 MuiButton-fullWidth-76"
      data-cy="submit"
      disabled={false}
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="submit"
    >
      <span
        className="MuiButton-label-52"
      >
        Anmelden
      </span>
      <span
        className="MuiTouchRipple-root-80"
      />
    </button>
  </form>
</main>
`;

exports[`components LoginForm renders error message if login failed 1`] = `
<main
  className="LoginForm-layout-1"
>
  <p
    className="MuiTypography-root-87 MuiTypography-body1-96 LoginForm-failureMessage-2"
  >
    Falsche Login-Daten
  </p>
  <form
    onSubmit={[Function]}
  >
    <div
      className="MuiFormControl-root-3 MuiFormControl-marginNormal-4 MuiFormControl-fullWidth-6"
    >
      <label
        className="MuiFormLabel-root-14 MuiFormLabel-filled-18 MuiFormLabel-required-19 MuiInputLabel-root-7 MuiInputLabel-formControl-8 MuiInputLabel-animated-11 MuiInputLabel-shrink-10"
        data-shrink={true}
        htmlFor="email"
      >
        E-Mail
        <span
          className="MuiFormLabel-asterisk-20"
        >
           *
        </span>
      </label>
      <div
        className="MuiInputBase-root-34 MuiInput-root-21 MuiInput-underline-25 MuiInputBase-formControl-35 MuiInput-formControl-22"
        data-cy="email"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoComplete="email"
          autoFocus={true}
          className="MuiInputBase-input-44 MuiInput-input-29"
          disabled={false}
          id="email"
          name="email"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={true}
          type="text"
          value="test@example.com"
        />
      </div>
    </div>
    <div
      className="MuiFormControl-root-3 MuiFormControl-marginNormal-4 MuiFormControl-fullWidth-6"
    >
      <label
        className="MuiFormLabel-root-14 MuiFormLabel-required-19 MuiInputLabel-root-7 MuiInputLabel-formControl-8 MuiInputLabel-animated-11"
        data-shrink={false}
        htmlFor="password"
      >
        Passwort
        <span
          className="MuiFormLabel-asterisk-20"
        >
           *
        </span>
      </label>
      <div
        className="MuiInputBase-root-34 MuiInput-root-21 MuiInput-underline-25 MuiInputBase-formControl-35 MuiInput-formControl-22"
        data-cy="password"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoComplete="current-password"
          className="MuiInputBase-input-44 MuiInput-input-29 MuiInputBase-inputType-47 MuiInput-inputType-32"
          disabled={false}
          id="password"
          name="password"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={true}
          type="password"
          value=""
        />
      </div>
    </div>
    <button
      className="MuiButtonBase-root-77 MuiButton-root-51 MuiButton-contained-62 MuiButton-containedPrimary-63 MuiButton-raised-65 MuiButton-raisedPrimary-66 MuiButton-fullWidth-76"
      data-cy="submit"
      disabled={false}
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="submit"
    >
      <span
        className="MuiButton-label-52"
      >
        Anmelden
      </span>
      <span
        className="MuiTouchRipple-root-80"
      />
    </button>
  </form>
</main>
`;
